vault/plugin_catalog.go

```diff
@@ -41,78 +42,119 @@ type PluginCatalog struct {
 	logger          log.Logger
 
 	// multiplexedClients holds plugin process connections by plugin name
-	// This allows a single grpc connection to communicate with multiple
-	// databases. Each database configuration using the same plugin will be
-	// routed to the existing plugin process.
+	//
+	// This allows plugins that suppport multiplexing to use a single grpc
+	// connection to communicate with multiple "backends". Each backend
+	// configuration using the same plugin will be routed to the existing
+	// plugin process.
 	multiplexedClients map[string]*MultiplexedClient
 
 	lock sync.RWMutex
 }
 
+// PluginClient represents a connection to a plugin process
 type PluginClient struct {
-	logger   log.Logger
-	id       string
-	protocol plugin.ClientProtocol
+	logger log.Logger
+
+	// id is the connection ID
+	id string
+
 	// client handles the lifecycle of a plugin process
-	client *plugin.Client
+	// multiplexed plugins share the same client
+	client      *plugin.Client
+	protocol    *pluginClientConn
+	cleanupFunc func() error
 }
 
+// MultiplexedClient holds client connections for multiplexed and
+// non-multiplexed plugin processes
 type MultiplexedClient struct {
 	logger log.Logger
 
 	// name is the plugin name
 	name string
 
-	client *plugin.Client
-
-	connections         map[string]*PluginClient
-	multiplexingSupport bool
+	// connections holds client connections by ID
+	connections map[string]*PluginClient
 }
 
-func (p *PluginClient) Conn() *grpc.ClientConn {
-	gc := p.protocol.(*plugin.GRPCClient)
-	return gc.Conn
+func (c *Core) setupPluginCatalog(ctx context.Context) error {
+	c.pluginCatalog = &PluginCatalog{
+		builtinRegistry: c.builtinRegistry,
+		catalogView:     NewBarrierView(c.barrier, pluginCatalogPath),
+		directory:       c.pluginDirectory,
+		logger:          c.logger,
+	}
+
+	// Run upgrade if untyped plugins exist
+	err := c.pluginCatalog.UpgradePlugins(ctx, c.logger)
+	if err != nil {
+		c.logger.Error("error while upgrading plugin storage", "error", err)
+		return err
+	}
+
+	if c.logger.IsInfo() {
+		c.logger.Info("successfully setup plugin catalog", "plugin-directory", c.pluginDirectory)
+	}
+
+	return nil
 }
 
-func (p *PluginClient) ID() string {
-	return p.id
+type pluginClientConn struct {
```

Comment:
Would it make sense to have `PluginClient` implement `the grpc.ClientConnInterface` and remove this wrapper struct entirely?

Reply:
It'd probably make it too leaky, so maybe not.